onnxruntime/core/framework/memcpy.cc

```diff
@@ -11,16 +11,33 @@ Memcpy::Memcpy(const OpKernelInfo& info)
 }
 
 Status Memcpy::Compute(OpKernelContext* ctx) const {
-  const auto* X = ctx->Input<Tensor>(0);
-  Tensor* Y = ctx->Output(0, X->Shape());
-  Status retval = Info().GetDataTransferManager().CopyTensor(*X, *Y, Info().GetKernelDef().ExecQueueId());
+  Status retval;
+  MLDataType input_type_0 = ctx->InputType(0);
+  if (input_type_0->IsTensorType()) {
+    const auto* X = ctx->Input<Tensor>(0);
+    Tensor* Y = ctx->Output(0, X->Shape());
+    retval = Info().GetDataTransferManager().CopyTensor(*X, *Y, Info().GetKernelDef().ExecQueueId());
 
-  if (!retval.IsOK()) {
-    LOGS(ctx->Logger(), ERROR) << MakeString(retval.ErrorMessage(),
-                                             " Copying ", Node().InputDefs()[0]->Name(),
-                                             " to ", Node().OutputDefs()[0]->Name(),
-                                             " Input shape:", X->Shape(), " Output shape:", Y->Shape(),
-                                             " X data:", X->DataRaw(), " Y data:", Y->DataRaw());
+    if (!retval.IsOK()) {
+      // XXX: Can LOGS really log something that is on device?
+      LOGS(ctx->Logger(), ERROR) << MakeString(retval.ErrorMessage(),
+                                               " Copying ", Node().InputDefs()[0]->Name(),
+                                               " to ", Node().OutputDefs()[0]->Name(),
+                                               " Input shape:", X->Shape(), " Output shape:", Y->Shape(),
+                                               " X data:", X->DataRaw(), " Y data:", Y->DataRaw());
+    }
+  } else if (input_type_0->IsSparseTensorType()) {
+    const auto* X = ctx->Input<SparseTensor>(0);
```

Comment:
>const auto* X = ctx->Input<SparseTensor>(0 [](http://example.com/codeflow?start=4&length=42)

Consider adding metadata into kernelDef that would describe the format that the destination kernel accepts and combine format conversion with memcpy.

Reply:
Will be done in another PR

Reply:
Next PR